contracts/PreOrder.sol

```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./BatchNFT.sol";
+
+contract PreOrder is BatchNFT {
+    struct Participant {
+        address sender;
+        uint256 amount;
+        uint256 time;
+    }
+
+    // the key `_nextPreOrder` is incremented with each additional pre-order
+    // make sure `_nextPreOrder` starts from 1 instead of 0
+    mapping(address => uint256) private _preOrdered;
+    mapping(uint256 => Participant) private _preOrders;
+    uint256 private _nextPreOrder;
```

Comment:
Shall we use `@openzeppelin/contracts/utils/Counters.sol` for the incrementation here?

Reply:
updated to use the `Counter` utility.